File: helpers/python/sandbox_net.py
#!/usr/bin/env python3
"""cr-019: sandbox 受控出站 helper。

读环境变量 SANDBOX_PROXY_SOCK(SOCKS5h over UDS 代理),把 HTTP 请求经代理转发。
纯标准库,零三方依赖,适合被沙箱化任务 import。
"""
import os
import socket
import ssl
import sys
from urllib.parse import urlparse


def _socks5h_connect(proxy_sock: str, host: str, port: int) -> socket.socket:
    s = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
    s.connect(proxy_sock)
    # 问候:NO-AUTH
    s.sendall(b"\x05\x01\x00")
    assert s.recv(2) == b"\x05\x00", "代理拒绝 NO-AUTH"
    # 请求 CONNECT(DOMAIN ATYP,远程 DNS)
    hb = host.encode()
    s.sendall(b"\x05\x01\x00\x03" + bytes([len(hb)]) + hb + port.to_bytes(2, "big"))
    rep = s.recv(10)
    if len(rep) < 2 or rep[1] != 0:
        raise ConnectionError(f"SOCKS5 拒绝(reply={rep[1] if len(rep) > 1 else '?'})")
    return s


class Response:
    def __init__(self, status: int, headers: str, body: bytes):
        self.status = status
        self.headers = headers
        self.body = body

    def __repr__(self):
        return f"<Response {self.status}>"


def request(method: str, url: str, body=None, headers=None) -> Response:
    u = urlparse(url)
    host = u.hostname
    port = u.port or (443 if u.scheme == "https" else 80)
    path = u.path or "/"
    if u.query:
        path += "?" + u.query

    proxy = os.environ.get("SANDBOX_PROXY_SOCK")
    if not proxy:
        raise RuntimeError("SANDBOX_PROXY_SOCK 未设置:该 profile 无出站白名单")

    s = _socks5h_connect(proxy, host, port)
    if u.scheme == "https":
        ctx = ssl.create_default_context()
        s = ctx.wrap_socket(s, server_hostname=host)

    hdrs = {"Host": host, "Connection": "close"}
    if headers:
        hdrs.update(headers)
    req_line = f"{method} {path} HTTP/1.1\r\n"
    for k, v in hdrs.items():
        req_line += f"{k}: {v}\r\n"
    req_line += "\r\n"
    s.sendall(req_line.encode() + (body or b""))

    raw = b""
    while True:
        chunk = s.recv(65536)
        if not chunk:
            break
        raw += chunk
    head, _, resp_body = raw.partition(b"\r\n\r\n")
    head_str = head.decode("latin1")
    status_line = head_str.split("\r\n", 1)[0]
    status = int(status_line.split()[1])
    return Response(status, head_str, resp_body)


def get(url: str, headers=None) -> Response:
    return request("GET", url, headers=headers)


if __name__ == "__main__":
    # CLI: python3 sandbox_net.py GET <url>
    if len(sys.argv) >= 3:
        print(get(sys.argv[2]).body.decode("utf-8", "replace"))

File: helpers/node/sandbox-net.js
#!/usr/bin/env node
// cr-019: sandbox 受控出站 helper(SOCKS5h over UDS + 原始 HTTP/HTTPS)。
// 纯 Node 内置模块,零三方依赖。手法对齐 python helper:SOCKS5h 握手后在
// relay 流上手写 HTTP,避免 http.agent 复用已连接 socket 的关闭问题。

const net = require("net");
const tls = require("tls");
const { URL } = require("url");

function socks5hConnect(proxyPath, host, port, cb) {
  const s = net.createConnection(proxyPath);
  let state = 0; // 0=等问候回复, 1=等 CONNECT 回复
  s.once("connect", () => {
    s.write(Buffer.from([0x05, 0x01, 0x00])); // VER, NMETHODS=1, NO-AUTH
  });
  s.on("data", (chunk) => {
    if (state === 0) {
      if (chunk[0] !== 0x05 || chunk[1] !== 0x00) {
        s.destroy();
        return cb(new Error("proxy auth failed"));
      }
      const hb = Buffer.from(host);
      const req = Buffer.concat([
        Buffer.from([0x05, 0x01, 0x00, 0x03, hb.length]), hb,
        Buffer.from([port >> 8, port & 0xff]),
      ]);
      s.write(req);
      state = 1;
    } else if (state === 1) {
      if (chunk[1] !== 0x00) {
        s.destroy();
        return cb(new Error("socks5 reply " + chunk[1]));
      }
      // 关键:移除握手监听,让后续 HTTP 数据流由 rawHttp 接管
      s.removeAllListeners("data");
      cb(null, s);
    }
  });
  s.on("error", (e) => cb(e));
}

// 在已就绪的流上写原始 HTTP 请求,收集响应直到 EOF,解析状态码 + body。
function rawHttp(stream, requestStr, cb) {
  const chunks = [];
  stream.on("data", (c) => chunks.push(c));
  stream.on("end", () => {
    const raw = Buffer.concat(chunks);
    const sep = raw.indexOf("\r\n\r\n");
    const head = (sep >= 0 ? raw.slice(0, sep) : raw).toString("latin1");
    const statusLine = head.split("\r\n")[0];
    const status = parseInt(statusLine.split(" ")[1], 10);
    cb(null, { status, headers: head, body: sep >= 0 ? raw.slice(sep + 4) : Buffer.alloc(0) });
  });
  stream.on("error", (e) => cb(e));
  stream.write(requestStr);
}

function request(method, urlStr, body, headers, cb) {
  const u = new URL(urlStr);
  const host = u.hostname;
  const port = Number(u.port) || (u.protocol === "https:" ? 443 : 80);
  const path = (u.pathname || "/") + u.search;
  const proxy = process.env.SANDBOX_PROXY_SOCK;
  if (!proxy) return cb(new Error("SANDBOX_PROXY_SOCK 未设置"));

  const lines = [`${method} ${path} HTTP/1.1`, `Host: ${host}`, "Connection: close"];
  if (headers) for (const k in headers) lines.push(`${k}: ${headers[k]}`);
  const requestStr = lines.join("\r\n") + "\r\n\r\n" + (body || "");

  socks5hConnect(proxy, host, port, (err, sock) => {
    if (err) return cb(err);
    if (u.protocol === "https:") {
      const ts = tls.connect({ socket: sock, servername: host }, () => rawHttp(ts, requestStr, cb));
      ts.on("error", (e) => cb(e));
    } else {
      rawHttp(sock, requestStr, cb);
    }
  });
}

module.exports = { request, get: (u, cb) => request("GET", u, null, null, cb) };

if (require.main === module) {
  const [, , m, urlStr] = process.argv;
  request(m || "GET", urlStr, null, null, (err, r) => {
    if (err) { console.error(err.message); process.exit(1); }
    process.stdout.write(r.body);
  });
}
